counter.go

```diff
@@ -0,0 +1,127 @@
+package signalfx
+
+import (
+	"math"
+	"sync/atomic"
+	"time"
+)
+
+// A Counter represents a Counter metric (i.e., its internal value is
+// reset to zero by its PostReportHook, although this should not be
+// meaningful to client code).
+type Counter struct {
+	metric     string
+	dimensions map[string]string
+	value      uint64
+}
+
+// NewCounter returns a new Counter with the specified parameters.  It
+// does not copy the dimensions; client code should take care not to
+// modify them in a goroutine-unsafe manner.
+func NewCounter(
+	metric string,
+	dimensions map[string]string,
+	value uint64,
+) *Counter {
+	return &Counter{metric: metric, dimensions: dimensions, value: value}
+}
+
+// Inc increments a Counter's internal state.  It is goroutine-safe.
+func (c *Counter) Inc(delta uint64) uint64 {
+	return atomic.AddUint64(&c.value, delta)
+}
+
+// DataPoint returns a DataPoint reflecting the Counter's internal
+// state.  It does not reset that state, leaving that to
+// PostReportHook.
+func (c *Counter) DataPoint() *DataPoint {
+	value := atomic.LoadUint64(&c.value)
+	if value == 0 || value > math.MaxInt64 {
+		return nil
+	}
+	return &DataPoint{
+		Metric:     c.metric,
+		Timestamp:  time.Now(),
+		Type:       CounterType,
+		Dimensions: c.dimensions,
+		Value:      int64(value),
+	}
+}
+
+// PostReportHook resets a Counter's internal state.  It does this
+// intelligently, by subtracting the successfully-reported value from
+// the state (this way, the Counter may have been incremented in
+// between the create of the reported DataPoint and the reset, and the
+// state will be consistent).  Calling PostReportHook with a negative
+// value will result in a panic: counters may never take negative
+// values.
+func (c *Counter) PostReportHook(v int64) {
+	if v < 0 {
+		panic("negative counter should be impossible")
```

Comment:
I was just reading into gRPC server code and found something that might be similar use case for the counter issue 	https://github.com/grpc/grpc-go/blob/master/server.go#L239-L247
where they exit the program instead of panic, may be that seems a better way of handling where the program exits, this way it seems that there was nothing to recover from by the user since the system is in corrupted state so it logged and stopped.